Reduced sim mode solver

dbochkov-flexcompute opened the pull request:
Implements https://github.com/flexcompute/tidy3d/issues/1191. That is, there is a new simulation method `.subsection(region: Box, ...)` that can be used to create a smaller simulation based on original one. 
- Most of simulation fields are inherited, but could be optionally overridden
- Grid can be requested to be the same as in the parent simulation (`grid_spec="identical"`). In this case it's converted into a custom grid and the requested region is expanded to match parent grid lines. One small issue here is that if the parent grid already matches the requested region, than due to numerical precision one additional cell might be included. But that doesn't seem cause any real issues.
- Similarly, due to numerical precision issues, even when child grid is essentially the same as the parent grid, some of structures might become slightly misaligned with grid lines. Most prominently it shows when calling `ModeSolver._solver_eps()` of local mode solver. That is, effective boundaries of structures might differ by one pixel between parent and child simulations.
- Regarding inheriting symmetry: we discussed in https://github.com/flexcompute/tidy3d/issues/1191 that probably we can just not allow selecting a not compatible region which is not centered around symmetry plane. But I realized this could happen in the background of a remote mode solver call and that would cause an error. So, I implemented it in a way so that an incompatible region is automatically expanded to be centered around a symmetry plane.
- `.subsection(region: Box, ...)` is used internally in mode solver web api call and whether it is used or not is controlled by flag `reduce_simulation`. I set it by default to `True`, that is, all mode solver calls will use a reduced simulation. We can change it either to `False`, or make it so that only simulation containing custom mediums are reduced and also display a warning/info notifying that this is done. What do you think?
- This does significantly help to reduce data transfer for remote mode solver. For example, in this notebook https://github.com/flexcompute-readthedocs/tidy3d-docs/blob/096b84d99d301de4e3690fb6b4aa373064a2bb91/docs/source/notebooks/ThermallyTunedRingResonator.ipynb I couldn't really use remote mode solver because each solve would require around 500mb of upload, now it's less than 1mb. I also expect it to be useful for regular optic simulations in the context of heat solver. I do use it in the same notebook to remove irrelevant structures from optic simulation, but I anticipate more impactful usage eventually. (in this notebook, we are already restricted to a single component. Imagine performing heat simulation on a large scale circuit and then use `.subsection()` to  cut out individual components for optic simulations)
- On basic level, this feature extends method `.sel_inside()` from `SpatialDataArray` to other spatial data arrays (`ScalarFieldDataArray`, `ScalarFieldTimeDataArray`), custom medium classes and space time modulation classes. I realize that the name I used `.sel_inside(bounds: Bound)` might be a bit misleading, because we extract data not just inside but that covers `bounds`. An alternative could be `.reduce_to_cover(bounds: Bound)`, `.sel_to_cover(bounds: Bound)`. Any other suggestions?

I put three of you as reviewers, but I don't think all of you need to review the entire thing. Maybe @weiliangjin2021 can looks through changes to custom medium classes and modulation classes, while @tylerflex and/or @momchil-flex can comment on overall implementation?

tylerflex commented on tidy3d/plugins/mode/mode_solver.py:
```diff
@@ -926,3 +951,48 @@ def _validate_modes_size(self):
+
+    @cached_property
+    def reduced_simulation_copy(self):
+        """Strip non-related structures, mediums, monitors, and sources from simulation object.
```
instead of "non-related" maybe something about "objects not used by the mode solver"? it wasn't immediately clear what `non-related` meant until I dug into things a little.

tylerflex commented on tidy3d/web/api/mode.py:
```diff
@@ -64,6 +65,9 @@ def run(
         Optional callback function called when uploading file with ``bytes_in_chunk`` as argument.
     progress_callback_download : Callable[[float], None] = None
         Optional callback function called when downloading file with ``bytes_in_chunk`` as argument.
+    reduce_simulation : bool = True
```
probably a good idea to do it this way.

tylerflex commented on tidy3d/components/simulation.py:
```diff
@@ -2984,3 +2986,198 @@ def from_scene(cls, scene: Scene, **kwargs) -> Simulation:
+        if remove_outside_custom_mediums:
+
+            # check for special treatment in case of PML
+            if isinstance(boundary_spec, (PML, StablePML, Absorber)):
```
I think these three are all instances of `AbsorberSpec`, which itself is a `BoundaryEdge`, not a `BoundarySpec`. something seems a little strange to me, would a `simulation.boundary_spec` ever be an instance of `PML`, etc?

dbochkov-flexcompute replied:
oh, good catch, this does seem wrong and I didn't have a test for that. Will fix it, thanks!

tylerflex commented on tests/test_plugins/test_mode_solver.py:
```diff
@@ -213,6 +213,22 @@ def verify_dtype(ms):
+def check_ms_reduction(ms):
+
+    ms_red = ms.reduced_simulation_copy
+    print(ms_red.simulation.symmetry)
```
do we need to print here? it's fine but our test outputs are already quite long.

tylerflex reviewed (commented):
Thanks @dbochkov-flexcompute I think the implementation makes a lot of sense. I just had a few minor comments about some of the conversions. This will be a nice feature to have in ways we haven't anticipated yet, I'm sure!

weiliangjin2021 commented on tidy3d/components/medium.py:
```diff
@@ -2486,6 +2615,37 @@ def loss_upper_bound(self) -> float:
+            CustomPoleResidue with reduced data.
+        """
+        if not self.eps_inf.does_cover(bounds=bounds):
+            log.warning("Eps inf spatial data array does not fully cover the requested region.")
```
Eps inf -> 'eps_inf'

weiliangjin2021 commented on tidy3d/components/medium.py:
```diff
@@ -2486,6 +2615,37 @@ def loss_upper_bound(self) -> float:
+        for pole, residue in self.poles:
+
+            if not pole.does_cover(bounds=bounds):
+                log.warning("Pole spatial data array does not fully cover the requested region.")
```
I wonder if it is necessary to warn at all. It is expected that some structure will be outside the requested region. It might raise tons of warning from a dispersive custom medium.

dbochkov-flexcompute replied:
actually, here `bounds` would represent intersection of region from `.subsection(region=region, ...)` and the bounding box of underlying structure or simulation itself (if background medium). So, if the original structure was put together correctly (that is, custom medium either covers it entirely or only one point is provided along a given dimension for constant extrapolation), then it won't produce any warnings

weiliangjin2021 commented on tidy3d/components/medium.py:
```diff
@@ -3843,6 +4167,9 @@ def eps_dataarray_freq(
 
+    def _sel_custom_data_inside(self, bounds: Bound):
+        return self
+
```
The reduction doesn't seem to apply to anistoropic custom medium?

dbochkov-flexcompute replied:
It does apply. It's just a proper `sel_inside()` is already inherited from `AnisotropicMedium` and we just need to neutralize abstract `_sel_custom_data_inside` inherited from `AbstractCustomMedium`

dbochkov-flexcompute commented:
addressed all comments, plus:
- shortened error/warning message here https://github.com/flexcompute/tidy3d/pull/1316/files#diff-8592632a38ecdcce2cea8cafbc92bd6e4314181e5e15bf07c57741ad4cd11f57R174 when an object is completely outside of simulation. Otherwise objects with custom medium produce lots of printed output.
- there was an error in docstring https://github.com/flexcompute/tidy3d/pull/1316/files#diff-227742ff635e6a1930bb06970d64a6f1f6cb94df41d53545959e00e16e5e7adeR986. Do we test now docstring code for correctness?
- after adding more tests ran into an issue that diffraction monitor validation would fail with an ambiguous message here https://github.com/flexcompute/tidy3d/pull/1316/files#diff-7360ac4bbe331087dbf8aa41d7163c3d1f0e95986f0488ce103cf24e4520049bR1238. I think this is related to the fact that `eps_model(frequency=...)` only works for a single frequency in case of custom mediums, so I changed validation in that case to only "central" frequency for now. But are diffraction monitors even allowed for custom medium?

momchil-flex commented:
> * after adding more tests ran into an issue that diffraction monitor validation would fail with an ambiguous message here https://github.com/flexcompute/tidy3d/pull/1316/files#diff-7360ac4bbe331087dbf8aa41d7163c3d1f0e95986f0488ce103cf24e4520049bR1238. I think this is related to the fact that `eps_model(frequency=...)` only works for a single frequency in case of custom mediums, so I changed validation in that case to only "central" frequency for now. But are diffraction monitors even allowed for custom medium?

This is a good question. I think there are two issues, both of which probably lie outside the scope of this PR.

- `eps_model` works for a list of frequencies in some cases, and not in others. I encountered this before as well, in fact I think I fixed it for anisotropic media, but it's still an issue for custom media. The method docstring actually suggests that only a `float` should be passed, but I believe there are many instances where we pass an array instead. Ideally we should make this consistent across all media, or it will keep popping up.
- We have a [validator](https://github.com/flexcompute/tidy3d/blob/2d6bd787686bd25fd371143f176ec5a0baa35b8d/tidy3d/components/simulation.py#L1073-L1081) that is *meant* to validate that a diffraction monitor is in a homogeneous medium. However, what it actually does is compute the list of mediums that the plane intersects, and make sure that there's only one element in that list. This does **not** work for custom medium, which is a single medium which may nevertheless still be inhomogeneous physically. This same issue applies to the validator for plane wave/gaussian beam sources, as well as, in some sense, to the [TFSF validator](def _validate_tfsf_structure_intersections) about structure intersections. I am not sure what's a good way to fix this: it seems like we should either prohibit custom medium, or, if custom medium found, actually sample the permittivity and check if it is homogeneous?
